Error: failed to start a llama runner

When I run 

    ollama run mistral

it downloads properly but then fails to run it, with the following error:

    Error: failed to start a llama runner

luser:~ $ sudo journalctl -u ollama
[sudo] password for luser: 
Oct 06 13:38:58 CX600 systemd[1]: Started Ollama Service.
Oct 06 13:38:58 CX600 ollama[1992]: Couldn't find '/usr/share/ollama/.ollama/id_ed25519'. Generating new private key.
Oct 06 13:38:58 CX600 ollama[1992]: Your new public key is:
Oct 06 13:38:58 CX600 ollama[1992]: 
Oct 06 13:38:58 CX600 ollama[1992]: 2023/10/06 13:38:58 images.go:996: total blobs: 0
Oct 06 13:38:58 CX600 ollama[1992]: 2023/10/06 13:38:58 images.go:1003: total unused blobs removed: 0
Oct 06 13:38:58 CX600 ollama[1992]: 2023/10/06 13:38:58 routes.go:572: Listening on 127.0.0.1:11434
Oct 06 13:41:53 CX600 ollama[1992]: [GIN] 2023/10/06 - 13:41:53 | 200 |      36.185µs |       127.0.0.1 | HEAD     "/"
Oct 06 13:41:53 CX600 ollama[1992]: [GIN] 2023/10/06 - 13:41:53 | 200 |     206.837µs |       127.0.0.1 | GET      "/api/>
Oct 06 13:41:56 CX600 ollama[1992]: 2023/10/06 13:41:56 images.go:1495: redirected to: https://dd20bb891979d25aebc8bec07b>
Oct 06 13:47:36 CX600 ollama[1992]: 2023/10/06 13:47:36 download.go:235: success getting sha256:8daa9615cce30c259a9555b1c>
Oct 06 13:47:38 CX600 ollama[1992]: 2023/10/06 13:47:38 images.go:1495: redirected to: https://dd20bb891979d25aebc8bec07b>
Oct 06 13:47:38 CX600 ollama[1992]: 2023/10/06 13:47:38 download.go:235: success getting sha256:8c17c2ebb0ea011be9981cc39>
Oct 06 13:47:39 CX600 ollama[1992]: 2023/10/06 13:47:39 images.go:1495: redirected to: https://dd20bb891979d25aebc8bec07b>
Oct 06 13:47:39 CX600 ollama[1992]: 2023/10/06 13:47:39 download.go:235: success getting sha256:7c23fb36d80141c4ab8cdbb61>
Oct 06 13:47:40 CX600 ollama[1992]: 2023/10/06 13:47:40 images.go:1495: redirected to: https://dd20bb891979d25aebc8bec07b>
Oct 06 13:47:40 CX600 ollama[1992]: 2023/10/06 13:47:40 download.go:235: success getting sha256:bec56154823a9d2956cf28f6c>
Oct 06 13:47:41 CX600 ollama[1992]: 2023/10/06 13:47:41 images.go:1495: redirected to: https://dd20bb891979d25aebc8bec07b>
Oct 06 13:47:42 CX600 ollama[1992]: 2023/10/06 13:47:42 download.go:235: success getting sha256:e35ab70a78c78ebbbc4d2e2ea>
Oct 06 13:47:42 CX600 ollama[1992]: 2023/10/06 13:47:42 images.go:1495: redirected to: https://dd20bb891979d25aebc8bec07b>
Oct 06 13:47:43 CX600 ollama[1992]: 2023/10/06 13:47:43 download.go:235: success getting sha256:09fe89200c09e3fa8b36e77da>
Oct 06 13:48:00 CX600 ollama[1992]: [GIN] 2023/10/06 - 13:48:00 | 200 |          6m7s |       127.0.0.1 | POST     "/api/>
Oct 06 13:48:00 CX600 ollama[1992]: 2023/10/06 13:48:00 llama.go:239: 4096 MiB VRAM available, loading up to 36 GPU layers
Oct 06 13:48:00 CX600 ollama[1992]: 2023/10/06 13:48:00 llama.go:313: starting llama runner
Oct 06 13:48:00 CX600 ollama[1992]: 2023/10/06 13:48:00 llama.go:349: waiting for llama runner to start responding
Oct 06 13:48:02 CX600 ollama[1992]: 2023/10/06 13:48:02 llama.go:323: llama runner exited with error: signal: illegal ins>
Oct 06 13:50:00 CX600 ollama[1992]: 2023/10/06 13:50:00 llama.go:330: error starting llama runner: llama runner did not s>
Oct 06 13:50:00 CX600 ollama[1992]: 2023/10/06 13:50:00 llama.go:313: starting llama runner
Oct 06 13:50:00 CX600 ollama[1992]: 2023/10/06 13:50:00 llama.go:349: waiting for llama runner to start responding
Oct 06 13:50:00 CX600 ollama[1992]: 2023/10/06 13:50:00 llama.go:323: llama runner exited with error: signal: illegal ins>
Oct 06 13:52:00 CX600 ollama[1992]: 2023/10/06 13:52:00 llama.go:330: error starting llama runner: llama runner did not s>
Oct 06 13:52:00 CX600 ollama[1992]: [GIN] 2023/10/06 - 13:52:00 | 500 |          4m0s |       127.0.0.1 | POST     "/api/>
Oct 06 13:52:13 CX600 ollama[1992]: [GIN] 2023/10/06 - 13:52:13 | 200 |      20.552µs |       127.0.0.1 | HEAD     "/"
Oct 06 13:52:13 CX600 ollama[1992]: [GIN] 2023/10/06 - 13:52:13 | 200 |     242.456µs |       127.0.0.1 | GET      "/api/>
Oct 06 13:52:13 CX600 ollama[1992]: 2023/10/06 13:52:13 llama.go:239: 4096 MiB VRAM available, loading up to 36 GPU layers
Oct 06 13:52:13 CX600 ollama[1992]: 2023/10/06 13:52:13 llama.go:313: starting llama runner
Oct 06 13:52:13 CX600 ollama[1992]: 2023/10/06 13:52:13 llama.go:349: waiting for llama runner to start responding
Oct 06 13:52:14 CX600 ollama[1992]: 2023/10/06 13:52:14 llama.go:323: llama runner exited with error: signal: illegal ins>
Oct 06 13:54:13 CX600 ollama[1992]: 2023/10/06 13:54:13 llama.go:330: error starting llama runner: llama runner did not s>
Oct 06 13:54:13 CX600 ollama[1992]: 2023/10/06 13:54:13 llama.go:313: starting llama runner
Oct 06 13:54:13 CX600 ollama[1992]: 2023/10/06 13:54:13 llama.go:349: waiting for llama runner to start responding
Oct 06 13:54:13 CX600 ollama[1992]: 2023/10/06 13:54:13 llama.go:323: llama runner exited with error: signal: illegal ins>
Oct 06 13:56:13 CX600 ollama[1992]: 2023/10/06 13:56:13 llama.go:330: error starting llama runner: llama runner did not s>
Oct 06 13:56:13 CX600 ollama[1992]: [GIN] 2023/10/06 - 13:56:13 | 500 |          4m0s |       127.0.0.1 | POST     "/api/>
Oct 06 14:37:57 CX600 ollama[1992]: [GIN] 2023/10/06 - 14:37:57 | 200 |      19.934µs |       127.0.0.1 | HEAD     "/"
Oct 06 14:37:59 CX600 ollama[1992]: [GIN] 2023/10/06 - 14:37:57 | 200 |     252.629µs |       127.0.0.1 | GET      "/api/>
Oct 06 14:54:42 CX600 ollama[1992]: [GIN] 2023/10/06 - 14:54:42 | 200 |      19.907µs |       127.0.0.1 | HEAD     "/"
Oct 06 14:54:42 CX600 ollama[1992]: [GIN] 2023/10/06 - 14:54:42 | 200 |     229.742µs |       127.0.0.1 | GET      "/api/>
Oct 06 14:54:45 CX600 ollama[1992]: 2023/10/06 14:54:45 images.go:1495: redirected to: https://dd20bb891979d25aebc8bec07b>
Oct 06 15:00:53 CX600 ollama[1992]: 2023/10/06 15:00:53 download.go:235: success getting sha256:3230a638a2da7f51833ddf0f5>
Oct 06 15:00:55 CX600 ollama[1992]: 2023/10/06 15:00:55 images.go:1495: redirected to: https://dd20bb891979d25aebc8bec07b>
Oct 06 15:00:55 CX600 ollama[1992]: 2023/10/06 15:00:55 download.go:235: success getting sha256:d5311aab7c4cecbb387fbb06d>
Oct 06 15:00:56 CX600 ollama[1992]: 2023/10/06 15:00:56 images.go:1495: redirected to: https://dd20bb891979d25aebc8bec07b>
Oct 06 15:00:57 CX600 ollama[1992]: 2023/10/06 15:00:57 download.go:235: success getting sha256:1e836a895a40bb08d7f5d3209>
Oct 06 15:01:16 CX600 ollama[1992]: [GIN] 2023/10/06 - 15:01:16 | 200 |         6m33s |       127.0.0.1 | POST     "/api/>
Oct 06 15:01:18 CX600 ollama[1992]: 2023/10/06 15:01:18 llama.go:239: 4096 MiB VRAM available, loading up to 32 GPU layers
Oct 06 15:01:18 CX600 ollama[1992]: 2023/10/06 15:01:18 llama.go:313: starting llama runner
Oct 06 15:01:18 CX600 ollama[1992]: 2023/10/06 15:01:18 llama.go:349: waiting for llama runner to start responding
Oct 06 15:01:20 CX600 ollama[1992]: 2023/10/06 15:01:20 llama.go:323: llama runner exited with error: signal: illegal ins>
Oct 06 15:03:18 CX600 ollama[1992]: 2023/10/06 15:03:18 llama.go:330: error starting llama runner: llama runner did not s>
Oct 06 15:03:18 CX600 ollama[1992]: 2023/10/06 15:03:18 llama.go:313: starting llama runner
Oct 06 15:03:18 CX600 ollama[1992]: 2023/10/06 15:03:18 llama.go:349: waiting for llama runner to start responding
Oct 06 15:03:18 CX600 ollama[1992]: 2023/10/06 15:03:18 llama.go:323: llama runner exited with error: signal: illegal ins>
Oct 06 15:05:18 CX600 ollama[1992]: 2023/10/06 15:05:18 llama.go:330: error starting llama runner: llama runner did not s>
Oct 06 15:05:18 CX600 ollama[1992]: [GIN] 2023/10/06 - 15:05:18 | 500 |          4m1s |       127.0.0.1 | POST     "/api/>
Oct 06 15:12:03 CX600 ollama[1992]: [GIN] 2023/10/06 - 15:12:03 | 200 |      19.418µs |       127.0.0.1 | HEAD     "/"
Oct 06 15:12:03 CX600 ollama[1992]: [GIN] 2023/10/06 - 15:12:03 | 200 |     268.109µs |       127.0.0.1 | GET      "/api/>
Oct 06 15:12:05 CX600 ollama[1992]: 2023/10/06 15:12:05 images.go:1495: redirected to: https://dd20bb891979d25aebc8bec07b>
Oct 06 15:14:56 CX600 ollama[1992]: 2023/10/06 15:14:56 download.go:235: success getting sha256:e84705205f71dd55be7b24a77>
Oct 06 15:14:58 CX600 ollama[1992]: 2023/10/06 15:14:58 images.go:1495: redirected to: https://dd20bb891979d25aebc8bec07b>
Oct 06 15:14:58 CX600 ollama[1992]: 2023/10/06 15:14:58 download.go:235: success getting sha256:e7214e2f1a0f5ed0ed67c3db9>
Oct 06 15:14:59 CX600 ollama[1992]: 2023/10/06 15:14:59 images.go:1495: redirected to: https://dd20bb891979d25aebc8bec07b>
Oct 06 15:14:59 CX600 ollama[1992]: 2023/10/06 15:14:59 download.go:235: success getting sha256:93ca9b3d83dc541f11062c0b9>
Oct 06 15:15:00 CX600 ollama[1992]: 2023/10/06 15:15:00 images.go:1495: redirected to: https://dd20bb891979d25aebc8bec07b>
Oct 06 15:15:00 CX600 ollama[1992]: 2023/10/06 15:15:00 download.go:235: success getting sha256:65009e4e7fee047467033b69d>
Oct 06 15:15:09 CX600 ollama[1992]: [GIN] 2023/10/06 - 15:15:09 | 200 |          3m5s |       127.0.0.1 | POST     "/api/>
Oct 06 15:15:09 CX600 ollama[1992]: 2023/10/06 15:15:09 llama.go:239: 4096 MiB VRAM available, loading up to 57 GPU layers
Oct 06 15:15:09 CX600 ollama[1992]: 2023/10/06 15:15:09 llama.go:313: starting llama runner
Oct 06 15:15:09 CX600 ollama[1992]: 2023/10/06 15:15:09 llama.go:349: waiting for llama runner to start responding
Oct 06 15:15:10 CX600 ollama[1992]: 2023/10/06 15:15:10 llama.go:323: llama runner exited with error: signal: illegal ins>
Oct 06 15:17:09 CX600 ollama[1992]: 2023/10/06 15:17:09 llama.go:330: error starting llama runner: llama runner did not s>
Oct 06 15:17:09 CX600 ollama[1992]: 2023/10/06 15:17:09 llama.go:313: starting llama runner
Oct 06 15:17:09 CX600 ollama[1992]: 2023/10/06 15:17:09 llama.go:349: waiting for llama runner to start responding
Oct 06 15:17:09 CX600 ollama[1992]: 2023/10/06 15:17:09 llama.go:323: llama runner exited with error: signal: illegal ins>
Oct 06 15:19:09 CX600 ollama[1992]: 2023/10/06 15:19:09 llama.go:330: error starting llama runner: llama runner did not s>
Oct 06 15:19:09 CX600 ollama[1992]: [GIN] 2023/10/06 - 15:19:09 | 500 |          4m0s |       127.0.0.1 | POST     "/api/>
Oct 06 15:54:16 CX600 ollama[1992]: [GIN] 2023/10/06 - 15:54:16 | 200 |      21.369µs |       127.0.0.1 | HEAD     "/"
Oct 06 15:54:18 CX600 ollama[1992]: [GIN] 2023/10/06 - 15:54:16 | 200 |     345.576µs |       127.0.0.1 | GET      "/api/>
Oct 06 15:54:18 CX600 ollama[1992]: 2023/10/06 15:54:16 llama.go:239: 4096 MiB VRAM available, loading up to 57 GPU layers
Oct 06 15:54:18 CX600 ollama[1992]: 2023/10/06 15:54:16 llama.go:313: starting llama runner
Oct 06 15:54:18 CX600 ollama[1992]: 2023/10/06 15:54:16 llama.go:349: waiting for llama runner to start responding
Oct 06 15:54:20 CX600 ollama[1992]: 2023/10/06 15:54:19 llama.go:323: llama runner exited with error: signal: illegal ins>
Oct 06 15:56:16 CX600 ollama[1992]: 2023/10/06 15:56:16 llama.go:330: error starting llama runner: llama runner did not s>
Oct 06 15:56:16 CX600 ollama[1992]: 2023/10/06 15:56:16 llama.go:313: starting llama runner
Oct 06 15:56:16 CX600 ollama[1992]: 2023/10/06 15:56:16 llama.go:349: waiting for llama runner to start responding
Oct 06 15:56:16 CX600 ollama[1992]: 2023/10/06 15:56:16 llama.go:323: llama runner exited with error: signal: illegal ins>
Oct 06 15:58:16 CX600 ollama[1992]: 2023/10/06 15:58:16 llama.go:330: error starting llama runner: llama runner did not s>
Oct 06 15:58:16 CX600 ollama[1992]: [GIN] 2023/10/06 - 15:58:16 | 500 |          4m0s |       127.0.0.1 | POST     "/api/>
luser:~ $

> * journalctl -u ollama

`Oct 12 03:01:30 TK5-100 ollama[80821]: 2023/10/12 03:01:30 llama.go:239: 16376 MiB VRAM available, loading up to 144 GPU layers
Oct 12 03:01:30 TK5-100 ollama[80821]: 2023/10/12 03:01:30 llama.go:313: starting llama runner
Oct 12 03:01:30 TK5-100 ollama[80821]: 2023/10/12 03:01:30 llama.go:349: waiting for llama runner to start responding
Oct 12 03:01:30 TK5-100 ollama[80821]: 2023/10/12 03:01:30 llama.go:323: llama runner exited with error: signal: illegal instruction (core dumped)
Oct 12 03:03:30 TK5-100 ollama[80821]: 2023/10/12 03:03:30 llama.go:330: error starting llama runner: llama runner did not start within alloted time, retrying
Oct 12 03:03:30 TK5-100 ollama[80821]: 2023/10/12 03:03:30 llama.go:313: starting llama runner
Oct 12 03:03:30 TK5-100 ollama[80821]: 2023/10/12 03:03:30 llama.go:349: waiting for llama runner to start responding
Oct 12 03:03:30 TK5-100 ollama[80821]: 2023/10/12 03:03:30 llama.go:323: llama runner exited with error: signal: illegal instruction (core dumped)
Oct 12 03:05:30 TK5-100 ollama[80821]: 2023/10/12 03:05:30 llama.go:330: error starting llama runner: llama runner did not start within alloted time, retrying
Oct 12 03:05:30 TK5-100 ollama[80821]: [GIN] 2023/10/12 - 03:05:30 | 500 |          4m0s |       127.0.0.1 | POST     "/api/generate"`

--- a/cmd/cmd.go
+++ b/cmd/cmd.go
@@ -479,18 +479,7 @@ func generate(cmd *cobra.Command, model, prompt string, wordWrap bool) error {
 	}
 
 	if err := client.Generate(cancelCtx, &request, fn); err != nil {
-		if strings.Contains(err.Error(), "failed to load model") {
-			// tell the user to check the server log, if it exists locally
-			home, nestedErr := os.UserHomeDir()
-			if nestedErr != nil {
-				// return the original error
-				return err
-			}
-			logPath := filepath.Join(home, ".ollama", "logs", "server.log")
-			if _, nestedErr := os.Stat(logPath); nestedErr == nil {
-				err = fmt.Errorf("%w\nFor more details, check the error logs at %s", err, logPath)
-			}
-		} else if strings.Contains(err.Error(), "context canceled") && abort {
+		if strings.Contains(err.Error(), "context canceled") && abort {
 			spinner.Finish()
 			return nil
 		}

--- a/llm/llama.go
+++ b/llm/llama.go
@@ -356,7 +356,10 @@ func newLlama(model string, adapters []string, runners []ModelRunner, numLayers 
 			log.Printf("error starting llama runner: %v", err)
 			llm.Close()
 
-			// capture the error runner process, if any
+			// default the runnerErr to the error returned by the most recent llama runner process
+			runnerErr = err
+
+			// capture the error directly from the runner process, if any
 			select {
 			case runnerErr = <-statusWriter.ErrCh:
 			default:
@@ -381,7 +384,7 @@ func newLlama(model string, adapters []string, runners []ModelRunner, numLayers 
 
 func waitForServer(llm *llama) error {
 	start := time.Now()
-	expiresAt := time.Now().Add(2 * time.Minute) // be generous with timeout, large models can take a while to load
+	expiresAt := time.Now().Add(3 * time.Minute) // be generous with timeout, large models can take a while to load
 	ticker := time.NewTicker(200 * time.Millisecond)
 	defer ticker.Stop()
 
@@ -394,7 +397,7 @@ func waitForServer(llm *llama) error {
 		case <-ticker.C:
 			if time.Now().After(expiresAt) {
 				// timeout
-				return fmt.Errorf("llama runner did not start within allotted time, retrying")
+				return fmt.Errorf("timed out waiting for llama runner to start")
 			}
 
 			if err := llm.Ping(context.Background()); err == nil {

--- a/llm/llama.go
+++ b/llm/llama.go
@@ -377,7 +377,6 @@ func newLlama(model string, adapters []string, runners []ModelRunner, numLayers 
 		return nil, runnerErr
 	}
 
-	// if you update this error message, also update the error check in cmd.generate()
 	return nil, fmt.Errorf("failed to start a llama runner")
 }
 

--- a/llm/llama.go
+++ b/llm/llama.go
@@ -263,7 +263,6 @@ func (w *StatusWriter) Write(b []byte) (int, error) {
 	if _, after, ok := bytes.Cut(b, []byte("error:")); ok {
 		err := fmt.Errorf("llama runner: %s", after)
 		w.ErrCh <- err
-		return os.Stderr.Write(b)
 	}
 	return os.Stderr.Write(b)
 }

--- a/llm/llama.go
+++ b/llm/llama.go
@@ -248,6 +248,26 @@ func NumGPU(numLayer, fileSizeBytes int64, opts api.Options) int {
 	return 1
 }
 
+// StatusWriter is a writer that captures error messages from the llama runner process
+type StatusWriter struct {
+	ErrCh chan error
+}
+
+func NewStatusWriter() *StatusWriter {
+	return &StatusWriter{
+		ErrCh: make(chan error, 1),
+	}
+}
+
+func (w *StatusWriter) Write(b []byte) (int, error) {
+	if _, after, ok := bytes.Cut(b, []byte("error:")); ok {
+		err := fmt.Errorf("llama runner: %s", after)
+		w.ErrCh <- err
+		return os.Stderr.Write(b)
+	}
+	return os.Stderr.Write(b)
+}
+
 func newLlama(model string, adapters []string, runners []ModelRunner, numLayers int64, opts api.Options) (*llama, error) {
 	fileInfo, err := os.Stat(model)
 	if err != nil {
@@ -294,6 +314,8 @@ func newLlama(model string, adapters []string, runners []ModelRunner, numLayers 
 		params = append(params, "--numa")
 	}
 
+	var runnerErr error
+
 	// start the llama.cpp server with a retry in case the port is already in use
 	for _, runner := range runners {
 		if _, err := os.Stat(runner.Path); err != nil {
@@ -310,7 +332,8 @@ func newLlama(model string, adapters []string, runners []ModelRunner, numLayers 
 		)
 		cmd.Env = append(os.Environ(), fmt.Sprintf("LD_LIBRARY_PATH=%s", filepath.Dir(runner.Path)))
 		cmd.Stdout = os.Stderr
-		cmd.Stderr = os.Stderr
+		statusWriter := NewStatusWriter()
+		cmd.Stderr = statusWriter
 
 		llm := &llama{Options: opts, Running: Running{Port: port, Cmd: cmd, Cancel: cancel, exitCh: make(chan error)}}
 
@@ -333,6 +356,14 @@ func newLlama(model string, adapters []string, runners []ModelRunner, numLayers 
 		if err := waitForServer(llm); err != nil {
 			log.Printf("error starting llama runner: %v", err)
 			llm.Close()
+
+			// capture the error runner process, if any
+			select {
+			case runnerErr = <-statusWriter.ErrCh:
+			default:
+				// the runner process probably timed out
+			}
+
 			// try again
 			continue
 		}
@@ -341,6 +372,12 @@ func newLlama(model string, adapters []string, runners []ModelRunner, numLayers 
 		return llm, nil
 	}
 
+	if runnerErr != nil {
+		// this is the error returned from the llama runner process that failed most recently
+		return nil, runnerErr
+	}
+
+	// if you update this error message, also update the error check in cmd.generate()
 	return nil, fmt.Errorf("failed to start a llama runner")
 }
 

